Sources/KSCrashRecording/Monitors/KSCrashMonitor_Memory.m

```diff
@@ -0,0 +1,837 @@
+ by apps.
+ 
+ We'll keep it private for now until it is integrated with `KSCrashMonitor_AppState`.
+ */
+@protocol AppStateTrackerObserving <NSObject>
+- (void)appStateTrackerDidChangeApplicationTransitionState:(KSCrash_ApplicationTransitionState)transitionState;
+@end
+
+typedef void (^AppStateTrackerBlockObserverBlock)(KSCrash_ApplicationTransitionState transitionState);
+@interface AppStateTrackerBlockObserver : NSObject <AppStateTrackerObserving>
+
+@property (nonatomic, copy) AppStateTrackerBlockObserverBlock block;
+@property (nonatomic, weak) id<AppStateTrackerObserving> object;
+
+- (BOOL)shouldReap;
+
+@end
+
+@implementation AppStateTrackerBlockObserver
+
+- (void)appStateTrackerDidChangeApplicationTransitionState:(KSCrash_ApplicationTransitionState)transitionState
+{
+    AppStateTrackerBlockObserverBlock block = self.block;
+    if (block) {
+        block(transitionState);
+    }
+    
+    id<AppStateTrackerObserving> object = self.object;
+    if (object) {
+        [object appStateTrackerDidChangeApplicationTransitionState:transitionState];
+    }
+}
+
+- (BOOL)shouldReap
+{
+    return self.block == nil && self.object == nil;
+}
+
+@end
+
+@interface AppStateTracker : NSObject {
+    
+    NSNotificationCenter *_center;
+    NSArray<id<NSObject>> *_registrations;
+    
+    // transition state and observers protected by the lock
+    os_unfair_lock _lock;
+    KSCrash_ApplicationTransitionState _transitionState;
+    NSMutableArray<id<AppStateTrackerObserving>> *_observers;
+}
+
+- (instancetype)initWithNotificationCenter:(NSNotificationCenter *)notificationCenter NS_DESIGNATED_INITIALIZER;
+- (instancetype)init NS_UNAVAILABLE;
++ (instancetype)new NS_UNAVAILABLE;
+
+@property (atomic, readonly) KSCrash_ApplicationTransitionState transitionState;
+
+- (void)addObserver:(id<AppStateTrackerObserving>)observer;
+- (id<AppStateTrackerObserving>)addObserverWithBlock:(AppStateTrackerBlockObserverBlock)block;
+
+- (void)removeObserver:(id<AppStateTrackerObserving>)observer;
+
+@end
+
+@implementation AppStateTracker
+
++ (void)load
+{
+    g_AppStateTracker = [[AppStateTracker alloc] initWithNotificationCenter:NSNotificationCenter.defaultCenter];
+    [g_AppStateTracker addObserverWithBlock:^(KSCrash_ApplicationTransitionState transitionState) {
+        _ks_memory_update(^(KSCrash_Memory *mem) {
+            mem->state = transitionState;
+        });
+    }];
+    [g_AppStateTracker start];
+}
+
+- (instancetype)initWithNotificationCenter:(NSNotificationCenter *)notificationCenter
+{
+    if (self = [super init]) {
+        _lock = OS_UNFAIR_LOCK_INIT;
+        _observers = [NSMutableArray array];
+        _center = notificationCenter;
+        _registrations = nil;
+        
+        BOOL isPrewarm = [NSProcessInfo.processInfo.environment[@"ActivePrewarm"] boolValue];
+        _transitionState = isPrewarm ? KSCrash_ApplicationTransitionStateStartup : KSCrash_ApplicationTransitionStateStartup;
```

Comment:
I think it should be like this
```suggestion
        _transitionState = isPrewarm ? KSCrash_ApplicationTransitionStateStartupPrewarm : KSCrash_ApplicationTransitionStateStartup;
```

Reply:
I think unit-test here are good to avoid this kind of bugs.

Reply:
Good catch, that would have made a difference in reporting an OOM or not. I'll tests for this.